Task BJS2-66096 completed:

Zhltsk-V opened the pull request:
Added the CommentController class containing methods for adding, updating, retrieving all, and deleting post comments

Added the CommentService class containing business logic for adding, updating, retrieving all, and deleting post comments

Added tests for both CommentController and CommentService classes

Added GlobalExceptionHandler for error handling

Commit:
Task BJS2-66096 completed:

Added the CommentController class containing methods for adding, updating, retrieving all, and deleting post comments

Added the CommentService class containing business logic for adding, updating, retrieving all, and deleting post comments

Added tests for both CommentController and CommentService classes

Added GlobalExceptionHandler for error handling

Commit:
Task BJS2-66096 completed:

Added the CommentController class containing methods for adding, updating, retrieving all, and deleting post comments

Added the CommentService class containing business logic for adding, updating, retrieving all, and deleting post comments

Added tests for both CommentController and CommentService classes

Added GlobalExceptionHandler for error handling

Commit:
Refactor tests and add Swagger/docs improvements

- Renamed test methods in CommentServiceTest and CommentControllerTest to follow Given-When-Then pattern
- Added comprehensive Swagger annotations (@Operation, @Schema) for:
  * All controller endpoints
  * Request/response DTOs (CommentCreateDto, CommentViewDto)
- Enhanced logging in controllers with English messages
- Maintained all existing test coverage while improving documentation

Commit:
Refactor tests and add Swagger/docs improvements

- Renamed test methods in CommentServiceTest and CommentControllerTest to follow Given-When-Then pattern
- Added comprehensive Swagger annotations (@Operation, @Schema) for:
  * All controller endpoints
  * Request/response DTOs (CommentCreateDto, CommentViewDto)
- Enhanced logging in controllers with English messages
- Maintained all existing test coverage while improving documentation

marsel-mkh commented on src/main/java/faang/school/postservice/controller/comment/CommentController.java:
```diff
@@ -0,0 +1,102 @@
+
+/**
+ * Контроллер для обработки запросов, связанных с комментариями.
+ * Предоставляет REST-эндпоинты для создания, обновления, получения и удаления комментариев.
```
Можно конкретно указать какие методы есть в классе с помощью аннотации @link

marsel-mkh commented on src/main/java/faang/school/postservice/controller/comment/CommentController.java:
```diff
@@ -0,0 +1,102 @@
+@Slf4j
+@RestController
+@RequiredArgsConstructor
+@RequestMapping("/posts/{postId}/comments")
```
Я не уверен, что так корректно указывать общий эндпоинт

marsel-mkh commented on src/main/java/faang/school/postservice/controller/comment/CommentController.java:
```diff
@@ -0,0 +1,102 @@
+            @PathVariable long postId,
+            @Parameter(description = "Comment data to create", required = true)
+            @RequestBody @Valid CommentCreateDto commentCreateDto) {
+        log.info("Request to create comment for post with ID: {}", postId);
```
можно добавить пробел, чтоб все не слипалось

marsel-mkh replied:
можно поправить везде

marsel-mkh commented on src/main/java/faang/school/postservice/controller/comment/GlobalExceptionHandler.java:
```diff
@@ -0,0 +1,78 @@
+     * @return Ответ с сообщением об ошибке и статусом HTTP 400 (Bad Request).
+     */
+    @ExceptionHandler(DataValidationException.class)
+    public ResponseEntity<String> handlerDataValidationException(DataValidationException e) {
```
Давай будем называть переменные словами, а не буквами, например exception

marsel-mkh replied:
поправь везде

marsel-mkh commented on src/main/java/faang/school/postservice/controller/comment/GlobalExceptionHandler.java:
```diff
@@ -0,0 +1,78 @@
+
+/**
+ * Глобальный обработчик исключений для всех контроллеров.
+ *
```
добавим какие методы есть в классе:?

marsel-mkh commented on src/main/java/faang/school/postservice/service/comment/CommentService.java:
```diff
@@ -0,0 +1,174 @@
+/**
+ * Сервис для работы с комментариями.
+ * Содержит бизнес-логику для создания, обновления, получения и удаления комментариев.
+ *
```
тут тоже можно добавить

marsel-mkh commented on src/main/java/faang/school/postservice/service/comment/CommentService.java:
```diff
@@ -0,0 +1,174 @@
+     * @throws DataValidationException Если комментарий не принадлежит указанному посту.
+     */
+    @Transactional
+    public CommentViewDto updateComment(Long postId, Long commentId, CommentCreateDto commentCreateDto) {
```
Если используешь @Transactional и достал сущность  внутри метода(почитай точно как работает), то сохранять ее не обязательно

marsel-mkh replied:
Если достал и добавил в нее что то, то она автоматом сохранит изменения

Zhltsk-V replied:
Аннотация не сохраняет данные автоматически, а обеспечивает: отслеживание изменений сущностей и фиксацию всех изменений при успешном завершении метода (если не было исключений).

marsel-mkh commented on src/main/java/faang/school/postservice/service/comment/CommentService.java:
```diff
@@ -0,0 +1,174 @@
+     * @return Найденный пост.
+     * @throws EntityNotFoundException Если пост не найден.
+     */
+    private Post getPostById(Long postId) {
```
этот метод есть у меня, если мое замержат, то можешь удалить

marsel-mkh commented on src/main/java/faang/school/postservice/service/comment/CommentService.java:
```diff
@@ -0,0 +1,174 @@
+     * @param commentId Идентификатор комментария (для сообщения об ошибке).
+     * @throws DataValidationException Если комментарий не принадлежит указанному посту.
+     */
+    private void validateCommentBelongsToPost(Comment comment, Long postId, Long commentId) {
```
Давай вынесем в отдельный слой валидации, так будет более правильно

marsel-mkh commented on src/test/java/faang/school/postservice/util/controller/comment/CommentControllerTest.java:
```diff
@@ -0,0 +1,218 @@
+@WebMvcTest(CommentController.class)
+public class CommentControllerTest {
+
+    @Autowired
```
Можно вынести  1 как валидный айди 999 как не валидный айди в константы, будет читабельнее

marsel-mkh commented on src/main/java/faang/school/postservice/service/comment/CommentService.java:
```diff
@@ -0,0 +1,174 @@
+        UserDto userDto = userServiceClient.getUser(authorId);
+        if (userDto == null) {
+            log.error("User with ID {} not found", authorId);
+            throw new EntityNotFoundException("User with ID " + authorId + " not found");
```
по идее это задвоение логов, в хэндлере тоже есть логи 
![image](https://github.com/user-attachments/assets/5da72d8e-1cb2-4bdc-b837-914e28f1daca)

marsel-mkh commented on src/main/java/faang/school/postservice/controller/comment/CommentController.java:
```diff
@@ -0,0 +1,102 @@
+            @PathVariable long postId) {
+        log.info("Request to get comments for post with ID: {}", postId);
+        List<CommentViewDto> comments = commentService.getCommentsByPostId(postId);
+        log.info("Found {} comments for post with ID: {}", comments.size(), postId);
```
Мне кажется или это слишком обильное покрытие логами?

Commit:
Merge branch 'murloc-master-stream9' into feature/BJS2-66096-comment-system

Commit:
Changes have been made to task BJS2-66096:

- Added the CommentValidator class and corresponding tests
- Fixed tests
- Removed the GlobalExceptionHandler (duplicate)

Commit:
Changes have been made to task BJS2-66096:

- Added the CommentValidator class and corresponding tests
- Fixed tests
- Removed the GlobalExceptionHandler (duplicate)

Commit:
This update includes several bugfixes

Commit:
This update includes several bugfixes

RinatBeybutov commented on src/main/java/faang/school/postservice/controller/comment/CommentController.java:
```diff
@@ -0,0 +1,110 @@
+    @PostMapping
+    public ResponseEntity<CommentViewDto> createComment(
+            @Parameter(description = "ID of the post to comment on", required = true, example = "1")
+            @PathVariable long postId,
```
давай поменяем на Long, лучше обертки использовать. Здесь и далее

RinatBeybutov commented on src/main/java/faang/school/postservice/mapper/CommentMapper.java:
```diff
@@ -0,0 +1,40 @@
+     * @param commentCreateDto DTO для создания комментария.
+     * @return Сущность Comment.
+     */
+    @Mapping(target = "id", ignore = true)
```
если в дто нет поля id, то оно по умолчанию будет игнорироваться, поэтому можно убрать эту строку. Проверь так же другие поля

RinatBeybutov commented on src/main/java/faang/school/postservice/validation/CommentValidator.java:
```diff
@@ -0,0 +1,58 @@
+     *
+     */
+    public void validateCommentBelongsToPost(Comment comment, Long postId, Long commentId) {
+        if (!comment.getPost().getId().equals(postId)) {
```
этот код мне не очень нравится, цепочка вызовов не самый удачный вариант из-за возможного null в цепочке. 
1. Можно на сущность Post навесить @EqualsAndHashCode и указать в ней только id
2. Либо можно руками прописать нужный метод сравнения в классе, чтобы немного инкапсулировать эту логику

RinatBeybutov commented on src/test/java/faang/school/postservice/service/LikeServiceTest.java:
```diff
@@ -78,15 +79,15 @@ public void givenValidPostAndUserWhenUnlikePostThenLikeIsDeleted() {
     @DisplayName("likeComment: позитивный сценарий")
```
здесь поправь, пожалуйста, что такое позитивный сценарий непонятно. Нужно как-то проще написать:
Проверка добавления лайка для комментария

Commit:
This update includes several bugfixes

Commit:
This update includes several bugfixes

Commit:
Merge branch 'murloc-master-stream9' into feature/BJS2-66096-comment-system

# Conflicts:
#	src/main/java/faang/school/postservice/controller/GlobalExceptionHandler.java

Commit:
Merge remote-tracking branch 'origin/feature/BJS2-66096-comment-system' into feature/BJS2-66096-comment-system

# Conflicts:
#	src/main/java/faang/school/postservice/controller/GlobalExceptionHandler.java

Commit:
This update includes several bugfixes
